cxx/emissions/sometimes.hh

```diff
@@ -1,50 +1,69 @@
 #pragma once
 
+#include <cmath>
 #include <unordered_map>
 
 #include "distributions/beta_bernoulli.hh"
 #include "emissions/base.hh"
 
-// An Emission class that sometimes applies BaseEmissor and sometimes doesn't.
-// BaseEmissor must assign zero probability to <clean, dirty> pairs with
-// clean == dirty.  [For example, BitFlip and Gaussian both satisfy this].
-template <typename BaseEmissor>
-class Sometimes : public Emission<typename std::tuple_element<
-                      0, typename BaseEmissor::SampleType>::type> {
+// An Emission class that sometimes applies another Emission and sometimes
+// doesn't.
+template <typename SampleType = double>
+class Sometimes : public Emission<SampleType> {
  public:
-  using SampleType =
-      typename std::tuple_element<0, typename BaseEmissor::SampleType>::type;
   BetaBernoulli bb;
-  BaseEmissor be;
+  Emission<SampleType>* be;  // We own be.
+  bool can_dirty_equal_clean;
 
-  Sometimes(){};
+  // By default, Sometimes only works with BaseEmissors that assign zero
+  // probability to dirty == clean.  Pass _can_dirty_equal_clean=true to
+  // override that assumption.
+  Sometimes(Emission<SampleType>* _be, bool _can_dirty_equal_clean = false):
+    be(_be), can_dirty_equal_clean(_can_dirty_equal_clean) {};
+
+  ~Sometimes() {
+    delete be;
+  }
 
   void incorporate(const std::pair<SampleType, SampleType>& x,
                    double weight = 1.0) {
     this->N += weight;
-    bb.incorporate(x.first != x.second, weight);
     if (x.first != x.second) {
-      be.incorporate(x, weight);
+      bb.incorporate(true, weight);
+      be->incorporate(x, weight);
+      return;
     }
+    if (can_dirty_equal_clean) {
+      double p_came_from_be = exp(bb.logp(true) + be->logp(x));
+      bb.incorporate(true, p_came_from_be * weight);
+      be->incorporate(x, p_came_from_be * weight);
+      bb.incorporate(false, (1.0 - p_came_from_be) * weight);
+      return;
+    }
+    bb.incorporate(false, weight);
   }
 
   double logp(const std::pair<SampleType, SampleType>& x) const {
     if (x.first != x.second) {
-      return bb.logp(true) + be.logp(x);
+      return bb.logp(true) + be->logp(x);
+    }
+    double lp = bb.logp(false);
+    if (can_dirty_equal_clean) {
+      lp += bb.logp(true) + be->logp(x);
```

Comment:
If we want `p_bb(~came_from_be) or (p_bb(came_from_be) and p_be(X=x))`, shouldn't this be `log(exp(bb.logp(false)) + exp(bb.logp(true) + be->logp(x)))`? (since the mutually exclusive probabilities combined by `or` should be added in probability space and not logp space)

Reply:
Good catch!  Done.